File: Dorkbox-Network/src/dorkbox/network/connection/PropertyStore.java
package dorkbox.network.connection;


import dorkbox.network.util.store.SettingsStore;
import dorkbox.util.SerializationManager;
import dorkbox.util.bytes.ByteArrayWrapper;
import dorkbox.util.database.DB_Server;
import dorkbox.util.database.DatabaseStorage;
import dorkbox.util.storage.Storage;
import dorkbox.util.storage.Store;
import org.bouncycastle.crypto.params.ECPrivateKeyParameters;
import org.bouncycastle.crypto.params.ECPublicKeyParameters;

import java.io.IOException;
import java.security.SecureRandom;
import java.util.HashMap;
import java.util.Map;

/**
 * The property store is the DEFAULT type of store for the network stack.
 * This is package private.
 */
public
class PropertyStore extends SettingsStore {

    protected Storage storage;
    protected Map<ByteArrayWrapper, DB_Server> servers;

    public
    PropertyStore() {
    }

    /**
     * Method of preference for creating/getting this connection store. package only since only the ConnectionStoreProxy calls this
     *
     * @param type                 this is either "Client" or "Server", depending on who is creating this endpoint.
     * @param serializationManager this is the serialization used for saving objects into the storage database
     */
    @Override
    public
    void init(Class<? extends EndPoint> type, final SerializationManager serializationManager, Storage storage) throws IOException {
        // make sure our custom types are registered
        serializationManager.register(HashMap.class);
        serializationManager.register(ByteArrayWrapper.class);
        serializationManager.register(DB_Server.class);

        if (storage == null) {
            this.storage = Store.Memory()
                                .make();
        }
        else {
            this.storage = storage;
        }

        servers = this.storage.load(DatabaseStorage.SERVERS, new HashMap<ByteArrayWrapper, DB_Server>(16));

        //use map to keep track of recid, so we can get record info during restarts.
        DB_Server localServer = servers.get(DB_Server.IP_0_0_0_0);
        if (localServer == null) {
            localServer = new DB_Server();
            servers.put(DB_Server.IP_0_0_0_0, localServer);

            // have to always specify what we are saving
            this.storage.commit(DatabaseStorage.SERVERS, servers);
        }
    }

    /**
     * Simple, property based method to getting the private key of the server
     */
    @Override
    public synchronized
    ECPrivateKeyParameters getPrivateKey() throws dorkbox.network.util.exceptions.SecurityException {
        checkAccess(EndPoint.class);

        return servers.get(DB_Server.IP_0_0_0_0)
                      .getPrivateKey();
    }

    /**
     * Simple, property based method for saving the private key of the server
     */
    @Override
    public synchronized
    void savePrivateKey(ECPrivateKeyParameters serverPrivateKey) throws dorkbox.network.util.exceptions.SecurityException {
        checkAccess(EndPoint.class);

        servers.get(DB_Server.IP_0_0_0_0)
               .setPrivateKey(serverPrivateKey);

        // have to always specify what we are saving
        storage.commit(DatabaseStorage.SERVERS, servers);
    }

    /**
     * Simple, property based method to getting the public key of the server
     */
    @Override
    public synchronized
    ECPublicKeyParameters getPublicKey() throws dorkbox.network.util.exceptions.SecurityException {
        checkAccess(EndPoint.class);

        return servers.get(DB_Server.IP_0_0_0_0)
                      .getPublicKey();
    }

    /**
     * Simple, property based method for saving the public key of the server
     */
    @Override
    public synchronized
    void savePublicKey(ECPublicKeyParameters serverPublicKey) throws dorkbox.network.util.exceptions.SecurityException {
        checkAccess(EndPoint.class);

        servers.get(DB_Server.IP_0_0_0_0)
               .setPublicKey(serverPublicKey);

        // have to always specify what we are saving
        storage.commit(DatabaseStorage.SERVERS, servers);
    }

    /**
     * Simple, property based method to getting the server salt
     */
    @Override
    public synchronized
    byte[] getSalt() {
        final DB_Server localServer = servers.get(DB_Server.IP_0_0_0_0);
        byte[] salt = localServer.getSalt();

        // we don't care who gets the server salt
        if (salt == null) {
            SecureRandom secureRandom = new SecureRandom();

            // server salt is used to salt usernames and other various connection handshake parameters
            byte[] bytes = new byte[256];
            secureRandom.nextBytes(bytes);

            salt = bytes;

            localServer.setSalt(bytes);

            // have to always specify what we are saving
            storage.commit(DatabaseStorage.SERVERS, servers);
        }

        return salt;
    }

    /**
     * Simple, property based method to getting a connected computer by host IP address
     */
    @Override
    public synchronized
    ECPublicKeyParameters getRegisteredServerKey(byte[] hostAddress) throws dorkbox.network.util.exceptions.SecurityException {
        checkAccess(RegistrationWrapper.class);

        final DB_Server db_server = this.servers.get(ByteArrayWrapper.wrap(hostAddress));
        if (db_server != null) {
            return db_server.getPublicKey();
        }
        return null;
    }

    /**
     * Saves a connected computer by host IP address and public key
     */
    @Override
    public synchronized
    void addRegisteredServerKey(byte[] hostAddress, ECPublicKeyParameters publicKey)
                    throws dorkbox.network.util.exceptions.SecurityException {
        checkAccess(RegistrationWrapper.class);

        final ByteArrayWrapper wrap = ByteArrayWrapper.wrap(hostAddress);
        DB_Server db_server = this.servers.get(wrap);
        if (db_server == null) {
            db_server = new DB_Server();
        }

        db_server.setPublicKey(publicKey);
        servers.put(wrap, db_server);
    }

    /**
     * Deletes a registered computer by host IP address
     */
    @Override
    public synchronized
    boolean removeRegisteredServerKey(byte[] hostAddress) throws dorkbox.network.util.exceptions.SecurityException {
        checkAccess(RegistrationWrapper.class);

        final ByteArrayWrapper wrap = ByteArrayWrapper.wrap(hostAddress);
        DB_Server db_server = this.servers.remove(wrap);

        return db_server != null;
    }

    @Override
    public
    void shutdown() {
        Store.close(storage);
    }
}

File: src/dorkbox/network/storage/PropertyStore.kt
/*
 * Copyright 2020 dorkbox, llc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dorkbox.network.storage

import dorkbox.netUtil.IPv4
import dorkbox.netUtil.IPv6
import dorkbox.network.connection.CryptoManagement
import dorkbox.network.serialization.Serialization
import dorkbox.util.storage.Storage
import org.agrona.collections.Object2NullableObjectHashMap
import java.net.InetAddress
import java.security.SecureRandom

/**
 * The property store is the DEFAULT type of store for the network stack.
 */
class PropertyStore : SettingsStore() {
    private lateinit var storage: Storage
    private lateinit var servers: Object2NullableObjectHashMap<InetAddress, DB_Server>

    /**
     * Address 0.0.0.0 or ::0 may be used as a source address for this host on this network.
     *
     * Because we assigned BOTH to the same thing, it doesn't matter which one we use
     */
    private val ipv4Host = IPv4.WILDCARD
    private val ipv6Host = IPv6.WILDCARD

    /**
     * Method of preference for creating/getting this connection store.
     *
     * @param serializationManager this is the serialization used for saving objects into the storage database
     */
    override fun init(serializationManager: Serialization, storage: Storage) {
        this.storage = storage
        servers = this.storage.get(DB_Server.STORAGE_KEY, Object2NullableObjectHashMap())

        // this will always be null and is here to help people that copy/paste code
        var localServer = servers[ipv4Host]
        if (localServer == null) {
            localServer = DB_Server()
            servers[ipv4Host] = localServer

            // have to always specify what we are saving
            this.storage.put(DB_Server.STORAGE_KEY, servers)
        }

        if (servers[ipv6Host] == null) {
            servers[ipv6Host] = localServer

            // have to always specify what we are saving
            this.storage.put(DB_Server.STORAGE_KEY, servers)
        }
    }

    /**
     * Simple, property based method to getting the private key of the server
     */
    @Synchronized
    override fun getPrivateKey(): ByteArray? {
        checkAccess(CryptoManagement::class.java)
        return servers[ipv4Host]!!.privateKey
    }

    /**
     * Simple, property based method for saving the private key of the server
     */
    @Synchronized
    override fun savePrivateKey(serverPrivateKey: ByteArray) {
        checkAccess(CryptoManagement::class.java)
        servers[ipv4Host]!!.privateKey = serverPrivateKey

        // have to always specify what we are saving
        storage.put(DB_Server.STORAGE_KEY, servers)
    }

    /**
     * Simple, property based method to getting the public key of the server
     */
    @Synchronized
    override fun getPublicKey(): ByteArray? {
        return servers[ipv4Host]!!.publicKey
    }

    /**
     * Simple, property based method for saving the public key of the server
     */
    @Synchronized
    override fun savePublicKey(serverPublicKey: ByteArray) {
        checkAccess(CryptoManagement::class.java)
        servers[ipv4Host]!!.publicKey = serverPublicKey

        // have to always specify what we are saving
        storage.put(DB_Server.STORAGE_KEY, servers)
    }

    /**
     * Simple, property based method to getting the server salt
     */
    @Synchronized
    override fun getSalt(): ByteArray {
        val localServer = servers[ipv4Host]
        var salt = localServer!!.salt

        // we don't care who gets the server salt
        if (salt == null) {
            val secureRandom = SecureRandom()

            // server salt is used to salt usernames and other various connection handshake parameters
            val bytes = ByteArray(256)
            secureRandom.nextBytes(bytes)
            salt = bytes
            localServer.salt = bytes

            // have to always specify what we are saving
            storage.put(DB_Server.STORAGE_KEY, servers)
        }

        return salt
    }

    /**
     * Simple, property based method to getting a connected computer by host IP address
     */
    @Synchronized
    override fun getRegisteredServerKey(hostAddress: InetAddress): ByteArray? {
        return servers[hostAddress]?.publicKey
    }

    /**
     * Saves a connected computer by host IP address and public key
     */
    @Synchronized
    override fun addRegisteredServerKey(hostAddress: InetAddress, publicKey: ByteArray) {
        // checkAccess(RegistrationWrapper.class);
        var db_server = servers[hostAddress]
        if (db_server == null) {
            db_server = DB_Server()
        }

        db_server.publicKey = publicKey
        servers[hostAddress] = db_server

        // have to always specify what we are saving
        storage.put(DB_Server.STORAGE_KEY, servers)
    }

    /**
     * Deletes a registered computer by host IP address
     */
    @Synchronized
    override fun removeRegisteredServerKey(hostAddress: InetAddress): Boolean {
        // checkAccess(RegistrationWrapper.class);
        val db_server = servers.remove(hostAddress)

        // have to always specify what we are saving
        storage.put(DB_Server.STORAGE_KEY, servers)

        return db_server != null
    }

    override fun close() {
        storage.close()
    }
}
